code/datums/elements/decal_element.dm

```diff
@@ -0,0 +1,178 @@
+	. = ..()
+	if(!isatom(target))
+		return ELEMENT_INCOMPATIBLE
+	if(_pic)
+		pic = _pic
+	else if(!generate_appearance(_icon, _icon_state, _dir, _layer, _color, _alpha, target))
+		return ELEMENT_INCOMPATIBLE
+	description = _description
+	cleanable = _cleanable
+	directional = _dir
+	base_icon_state = _icon_state
+
+	RegisterSignal(target, COMSIG_ATOM_UPDATE_OVERLAYS, PROC_REF(apply_overlay), TRUE)
+	if(target.initialized)
+		target.update_appearance(UPDATE_OVERLAYS) //could use some queuing here now maybe.
+	else
+		RegisterSignal(target, COMSIG_ATOM_AFTER_SUCCESSFUL_INITIALIZE, PROC_REF(late_update_icon), TRUE)
+	if(isitem(target))
+		INVOKE_ASYNC(target, TYPE_PROC_REF(/obj/item/, update_slot_icon), TRUE)
+	if(_dir)
+		RegisterSignal(target, COMSIG_ATOM_DECALS_ROTATING, PROC_REF(shuttle_rotate), TRUE)
+		SSdcs.RegisterSignal(target, COMSIG_ATOM_DIR_CHANGE, TYPE_PROC_REF(/datum/controller/subsystem/processing/dcs, rotate_decals), override=TRUE)
+	if(_cleanable)
+		RegisterSignal(target, COMSIG_COMPONENT_CLEAN_ACT, PROC_REF(clean_react), TRUE)
+	if(_description)
+		RegisterSignal(target, COMSIG_PARENT_EXAMINE, PROC_REF(examine),TRUE)
+
+	RegisterSignal(target, COMSIG_TURF_ON_SHUTTLE_MOVE, PROC_REF(shuttle_move_react),TRUE)
+
+
+/datum/element/decal/proc/late_update_icon(atom/source)
+	SIGNAL_HANDLER  // COMSIG_ATOM_AFTER_SUCCESSFUL_INITIALIZE
+
+	if(istype(source) && !(source.flags_2 & DECAL_INIT_UPDATE_EXPERIENCED_2))
+		source.flags_2 |= DECAL_INIT_UPDATE_EXPERIENCED_2
+		source.update_appearance(UPDATE_OVERLAYS)
+		UnregisterSignal(source, COMSIG_ATOM_AFTER_SUCCESSFUL_INITIALIZE)
+
+/**
+ * If the decal was not given an appearance, it will generate one based on the other given arguments.
+ * element won't be compatible if it cannot do either
+ * all args are fed into creating an image, they are byond vars for images you'll recognize in the byond docs
+ * (except source, source is the object whose appearance we're copying.)
+ */
+/datum/element/decal/proc/generate_appearance(_icon, _icon_state, _dir, _layer, _color, _alpha, source)
+	if(!_icon || !_icon_state)
+		return FALSE
+	var/temp_image = image(_icon, null, _icon_state, _layer, _dir)
+	pic = new(temp_image)
+	pic.color = _color
+	pic.alpha = _alpha
+	return TRUE
+
+/datum/element/decal/Detach(atom/source)
+	UnregisterSignal(source, list(
+		COMSIG_ATOM_DIR_CHANGE,
+		COMSIG_COMPONENT_CLEAN_ACT,
+		COMSIG_PARENT_EXAMINE,
+		COMSIG_ATOM_UPDATE_OVERLAYS,
+		COMSIG_TURF_ON_SHUTTLE_MOVE,
+		COMSIG_ATOM_DECALS_ROTATING
+	))
+	SSdcs.UnregisterSignal(source, COMSIG_ATOM_DIR_CHANGE)
+	source.update_appearance(UPDATE_OVERLAYS)
+	if(isitem(source))
+		INVOKE_ASYNC(source, TYPE_PROC_REF(/obj/item/, update_slot_icon))
+	return ..()
+
+/datum/element/decal/proc/apply_overlay(atom/source)
+	SIGNAL_HANDLER
+
+	source.add_overlay(pic)
+	// TODO: Fix this disgusting hack
```

Comment:
noting the todo here

Reply:
Yup. `left as an exercise for the next poor sap to touch this code (probably me)`